# 🚀 BHARATH BUILDS — 21 DAYS TO ML ENGINEER

## DAY 4 — REGULARIZATION + MULTIPLE REGRESSION

### 🎯 Objective

By the end of this practical, we will:

- Understand Multiple Linear Regression using multiple features.
- Train a baseline Linear Regression model.
- Understand how Ridge (L2) regularization controls model complexity.
- Understand how Lasso (L1) regularization can perform feature selection.
- Compare Linear Regression, Ridge, and Lasso using evaluation metrics.
- Compare model coefficients to understand the effect of regularization.
- Use Cross-Validation to compare model generalization.

---

## 📚 Today's Topics

### Whiteboard Concepts
- Multiple Linear Regression
- Underfitting vs Good Fit vs Overfitting
- Multicollinearity
- Regularization
- Regularization Parameter (Alpha)
- Ridge Regression — L2
- Lasso Regression — L1
- Ridge vs Lasso
- Feature Scaling
- Bias-Variance Trade-off
- Generalization
- Cross-Validation

### 💻 Practical Implementation

We will:

1. Load the California Housing dataset.
2. Perform only the minimum dataset checks required.
3. Split the data into training and testing sets.
4. Train Multiple Linear Regression.
5. Evaluate the baseline model.
6. Train Ridge Regression.
7. Evaluate Ridge.
8. Train Lasso Regression.
9. Evaluate Lasso.
10. Compare model coefficients.
11. Perform 5-Fold Cross-Validation.
12. Create a final Regularization Comparison Table.

---

## 🛠️ Tools Used

- Python
- Google Colab
- NumPy
- Pandas
- Scikit-learn
- Matplotlib

---

## 📊 Dataset

**California Housing Dataset**

The dataset contains multiple numerical features that we will use to predict a continuous target.

### Target:
House Value

### Models:
- Linear Regression
- Ridge Regression
- Lasso Regression

---

## 🎯 Main Goal

Today's practical question:

> **How does regularization affect model coefficients and generalization when we use multiple features?**

### Learning Flow

DATA
↓
MULTIPLE FEATURES
↓
TRAIN / TEST SPLIT
↓
LINEAR REGRESSION
↓
RIDGE (L2)
↓
LASSO (L1)
↓
COMPARE METRICS
↓
COMPARE COEFFICIENTS
↓
CROSS-VALIDATION
↓
CHECK GENERALIZATION

---

## 🚀 Expected Final Result

At the end of this notebook, we will create a:

### 📊 Regularization Comparison Table

| Model | Train R² | Test R² | Test MAE | Mean CV R² |
|---|---:|---:|---:|---:|
| Linear Regression | — | — | — | — |
| Ridge | — | — | — | — |
| Lasso | — | — | — | — |

We will also compare how the coefficients change between:

**Linear Regression → Ridge → Lasso**

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.datasets import fetch_california_housing
from sklearn.model_selection import train_test_split, cross_val_score

from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression, Ridge, Lasso

from sklearn.metrics import r2_score, mean_absolute_error,mean_squared_error

step1 : load the dataset

In [3]:
housing=fetch_california_housing(as_frame=True)
X=housing.data
y=housing.target

X.head()

,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude
0,8.3252,41.0,6.984127,1.023810,322.0,2.555556,37.88,-122.23
1,8.3014,21.0,6.238137,0.971880,2401.0,2.109842,37.86,-122.22
2,7.2574,52.0,8.288136,1.073446,496.0,2.802260,37.85,-122.24
3,5.6431,52.0,5.817352,1.073059,558.0,2.547945,37.85,-122.25
4,3.8462,52.0,6.281853,1.081081,565.0,2.181467,37.85,-122.25


step2: minimal dataset check EDA



In [4]:
print("shape",X.shape)
print("null values",X.isnull().sum())
print("\nFeatures:", X.columns)
print("\nTarget:", y.name)

shape (20640, 8)
null values MedInc        0
HouseAge      0
AveRooms      0
AveBedrms     0
Population    0
AveOccup      0
Latitude      0
Longitude     0
dtype: int64

Features: Index(['MedInc', 'HouseAge', 'AveRooms', 'AveBedrms', 'Population', 'AveOccup',
       'Latitude', 'Longitude'],
      dtype='object')

Target: MedHouseVal


step 3: train test split

In [6]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=42)

print("train shape",X_train.shape)
print("test shape",X_test.shape)

train shape (16512, 8)
test shape (4128, 8)


step4: train multiple linear regression

In [8]:
linear_model=LinearRegression()
linear_model.fit(X_train,y_train)

y_train_pred=linear_model.predict(X_train)
y_test_pred=linear_model.predict(X_test)

step5: evaluate linear regression

In [9]:
linear_train_r2=r2_score(y_train,y_train_pred)
linear_test_r2=r2_score(y_test,y_test_pred)

linear_mae=mean_absolute_error(y_test,y_test_pred)
linear_mse=mean_squared_error(y_test,y_test_pred)

print("linear train r2",linear_train_r2)
print("linear test r2",linear_test_r2)
print("linear mae",linear_mae)
print("linear mse",linear_mse)

linear train r2 0.6125511913966952
linear test r2 0.5757877060324508
linear mae 0.5332001304956553
linear mse 0.5558915986952444


step6 : inspect coefficients

In [10]:
linear_coefficients=pd.DataFrame(
    {
        "feature":X.columns,
        "coefficient":linear_model.coef_

    }
)

linear_coefficients.sort_values(by="coefficient",ascending=False,key=abs)

,feature,coefficient
3,AveBedrms,0.783145
0,MedInc,0.448675
7,Longitude,-0.433708
6,Latitude,-0.419792
2,AveRooms,-0.123323
1,HouseAge,0.009724
5,AveOccup,-0.003526
4,Population,-0.000002


step7: ridge regression

In [12]:
from sklearn.pipeline import Pipeline
ridge_model=Pipeline([
    ("scaler",StandardScaler()),
    ("ridge",Ridge())
])

ridge_model.fit(X_train,y_train)

ridge_train_pred=ridge_model.predict(X_train)
ridge_test_pred=ridge_model.predict(X_test)




step 8 : evaluate Ridge model

In [13]:
ridge_train_r2=r2_score(y_train,ridge_train_pred)
ridge_test_r2=r2_score(y_test,ridge_test_pred)

ridge_mae=mean_absolute_error(y_test,ridge_test_pred)
ridge_mse=mean_squared_error(y_test,ridge_test_pred)

print("ridge train r2",ridge_train_r2)
print("ridge test r2",ridge_test_r2)
print("ridge mae",ridge_mae)
print("ridge mse",ridge_mse)

ridge train r2 0.6125511218634558
ridge test r2 0.5758157428913684
ridge mae 0.5331931195789733
ridge mse 0.5558548589435971


step9 Lasso regression

In [14]:
lasso_model=Pipeline([
    ("scaler",StandardScaler()),
    ("lasso",Lasso())
])

lasso_model.fit(X_train,y_train)

lasso_train_pred=lasso_model.predict(X_train)
lasso_test_pred=lasso_model.predict(X_test)

step10: evaluate lasso model

In [15]:
lasso_train_r2=r2_score(y_train,lasso_train_pred)
lasso_test_r2=r2_score(y_test,lasso_test_pred)

lasso_mae=mean_absolute_error(y_test,lasso_test_pred)
lasso_mse=mean_squared_error(y_test,lasso_test_pred)

print("lasso train r2",lasso_train_r2)
print("lasso test r2",lasso_test_r2)
print("lasso mae",lasso_mae)
print("lasso mse",lasso_mse)

lasso train r2 0.0
lasso test r2 -0.00021908714592466794
lasso mae 0.9060685490007149
lasso mse 1.3106960720039365


step 11: compare all the model performance

In [17]:
results=pd.DataFrame({
    "model":["linear","ridge","lasso"],
    "train_r2":[linear_train_r2,ridge_train_r2,lasso_train_r2],
    "test_r2":[linear_test_r2,ridge_test_r2,lasso_test_r2],
    "mae":[linear_mae,ridge_mae,lasso_mae],
    "mse":[linear_mse,ridge_mse,lasso_mse],


})

results

,model,train_r2,test_r2,mae,mse
0,linear,0.612551,0.575788,0.533200,0.555892
1,ridge,0.612551,0.575816,0.533193,0.555855
2,lasso,0.000000,-0.000219,0.906069,1.310696


step 12 : compare coefficients

In [22]:
ridge_coefficients = pd.DataFrame(
    {
        "feature": X.columns,
        "coefficient": ridge_model["ridge"].coef_
    }
)

lasso_coefficients = pd.DataFrame(
    {
        "feature": X.columns,
        "coefficient": lasso_model["lasso"].coef_
    }
)

coefficient_comparison = pd.merge(
    linear_coefficients,
    ridge_coefficients,
    on="feature",
    suffixes=("_linear", "_ridge")
)

coefficient_comparison = pd.merge(
    coefficient_comparison,
    lasso_coefficients,
    on="feature"
)

coefficient_comparison = coefficient_comparison.rename(
    columns={
        "coefficient": "coefficient_lasso"
    }
)

coefficient_comparison

,feature,coefficient_linear,coefficient_ridge,coefficient_lasso
0,MedInc,0.448675,0.854327,0.0
1,HouseAge,0.009724,0.122624,0.0
2,AveRooms,-0.123323,-0.294210,0.0
3,AveBedrms,0.783145,0.339008,-0.0
4,Population,-0.000002,-0.002282,-0.0
5,AveOccup,-0.003526,-0.040833,-0.0
6,Latitude,-0.419792,-0.896168,-0.0
7,Longitude,-0.433708,-0.869071,-0.0


step 13: check lasso features selection

In [ ]:
lasso_features = coefficient_comparison[
    coefficient_comparison["coefficient_lasso"] != 0
]

lasso_features

In [35]:
lasso_features = coefficient_comparison[
    coefficient_comparison["coefficient_lasso"] != 0
]

lasso_features

,feature,coefficient_linear,coefficient_ridge,coefficient_lasso


step 14: cross validation step

In [38]:
models = {
    "Linear Regression": LinearRegression(),

    "Ridge": Pipeline([
        ("scaler", StandardScaler()),
        ("ridge", Ridge(alpha=1.0))
    ]),

    "Lasso": Pipeline([
        ("scaler", StandardScaler()),
        ("lasso", Lasso(alpha=0.01, max_iter=10000))
    ])
}

cv_results = []

for name, model in models.items():

    scores = cross_val_score(
        model,
        X_train,
        y_train,
        cv=5,
        scoring="r2"
    )

    cv_results.append({
        "Model": name,
        "Mean CV R²": scores.mean(),
        "Std CV R²": scores.std()
    })

cv_results = pd.DataFrame(cv_results)

cv_results

,Model,Mean CV R²,Std CV R²
0,Linear Regression,0.611484,0.006467
1,Ridge,0.611484,0.006460
2,Lasso,0.607711,0.004625


Step 15 — Final Regularization Comparison Table

In [39]:
final_comparison = pd.DataFrame({
    "Model": [
        "Linear Regression",
        "Ridge",
        "Lasso"
    ],
    "Train R²": [
        linear_train_r2,
        ridge_train_r2,
        lasso_train_r2
    ],
    "Test R²": [
        linear_test_r2,
        ridge_test_r2,
        lasso_test_r2
    ],
    "Test MAE": [
        linear_mae,
        ridge_mae,
        lasso_mae
    ]
})

final_comparison = final_comparison.merge(
    cv_results,
    on="Model"
)

final_comparison

,Model,Train R²,Test R²,Test MAE,Mean CV R²,Std CV R²
0,Linear Regression,0.612551,0.575788,0.533200,0.611484,0.006467
1,Ridge,0.612551,0.575816,0.533193,0.611484,0.006460
2,Lasso,0.000000,-0.000219,0.906069,0.607711,0.004625


## 📊 Results Interpretation

### Linear Regression
Linear Regression achieved a Train R² of 0.613 and Test R² of 0.576, giving us a reasonable baseline with no major overfitting.

### Ridge Regression
Ridge produced almost the same results as Linear Regression. The Test R² improved only slightly from 0.5758 to 0.5758, showing that regularization had very little effect at the selected alpha.

### Lasso Regression
Lasso performed much worse, with Train R² ≈ 0 and Test R² ≈ -0.0002. Its coefficients were pushed to zero, meaning the selected alpha was too strong and caused the model to underfit.

### Cross-Validation
Linear Regression and Ridge had almost identical mean CV R² scores (0.611), while Lasso was slightly lower (0.608).

### Key Insight
The main lesson is that regularization strength matters. Too much regularization can remove useful information and cause underfitting. Therefore, alpha should be tuned rather than chosen arbitrarily.

### Final Takeaway
Linear Regression provided the baseline, Ridge behaved very similarly, and the current Lasso configuration was too strongly regularized for this dataset.